hack/generated/pkg/testcommon/kube_test_context.go

```diff
@@ -22,57 +25,108 @@ import (
+// TODO: to genruntime. Same for errorAnnotation
+type KubeGlobalContext struct {
 	TestContext
-	KubeClient client.Client
-	Ensure     *Ensure
-	Match      *KubeMatcher
 
-	Namespace string
+	useEnvTest bool
+
+	namespace       string
+	stateAnnotation string
+	errorAnnotation string
 }
 
-// TODO: State Annotation parameter should be removed once the interface for Status determined and promoted
-// TODO: to genruntime. Same for errorAnnotation
-func NewKubeTestContext(
-	config *rest.Config,
-	region string,
+func (ctx KubeGlobalContext) Namespace() string {
+	return ctx.namespace
+}
+
+func NewKubeContext(
+	useEnvTest bool,
+	recordReplay bool,
 	namespace string,
+	region string,
 	stateAnnotation string,
-	errorAnnotation string) (*KubeTestContext, error) {
+	errorAnnotation string) KubeGlobalContext {
+	return KubeGlobalContext{
+		TestContext:     NewTestContext(region, recordReplay),
+		useEnvTest:      useEnvTest,
+		namespace:       namespace,
+		stateAnnotation: stateAnnotation,
+		errorAnnotation: errorAnnotation,
+	}
+}
+
+func (ctx KubeGlobalContext) ForTest(t *testing.T) (KubePerTestContext, error) {
+	perTestContext, err := ctx.TestContext.ForTest(t)
+	if err != nil {
+		return KubePerTestContext{}, err
+	}
 
-	clientOpts := client.Options{
-		Scheme: CreateScheme(),
+	var baseCtx *KubeBaseTestContext
+	if ctx.useEnvTest {
+		baseCtx, err = createEnvtestContext(perTestContext)
+	} else {
+		baseCtx, err = createRealKubeContext(perTestContext)
+	}
+
+	if err != nil {
+		return KubePerTestContext{}, err
 	}
 
-	kubeClient, err := client.New(config, clientOpts)
+	clientOptions := client.Options{Scheme: CreateScheme()}
+	kubeClient, err := client.New(baseCtx.KubeConfig, clientOptions)
 	if err != nil {
-		return nil, errors.Wrapf(err, "creating kubeclient")
+		return KubePerTestContext{}, err
 	}
 
-	ensure := NewEnsure(kubeClient, stateAnnotation, errorAnnotation)
+	ensure := NewEnsure(
+		kubeClient,
+		ctx.stateAnnotation,
+		ctx.errorAnnotation)
 
-	testContext, err := NewTestContext(region)
+	match := NewKubeMatcher(ensure)
+
+	result := KubePerTestContext{
+		KubeGlobalContext:   &ctx,
+		KubeBaseTestContext: *baseCtx,
+		KubeClient:          kubeClient,
+		Ensure:              ensure,
+		Match:               match,
+	}
+
+	err = result.createTestNamespace()
 	if err != nil {
-		return nil, err
+		return KubePerTestContext{}, err
 	}
 
-	return &KubeTestContext{
-		TestContext: *testContext,
-		Ensure:      ensure,
-		Match:       NewKubeMatcher(ensure),
-		KubeClient:  kubeClient,
-		Namespace:   namespace,
-	}, nil
+	return result, nil
+}
+
+type KubeBaseTestContext struct {
+	PerTestContext
+
+	KubeConfig *rest.Config
 }
 
-func (tc *KubeTestContext) CreateTestNamespace() error {
+type KubePerTestContext struct {
+	*KubeGlobalContext
+	KubeBaseTestContext
+
+	KubeClient client.Client
+	Ensure     *Ensure
+	Match      *KubeMatcher
+}
+
+func (tc *KubePerTestContext) createTestNamespace() error {
```

Comment:
inconsistent receivers -- ptr or not?

Reply:
originally this did mutation. it does not now. changed to non-ptr